packages/autocomplete-core/src/utils/createConcurrentSafePromise.ts

```diff
@@ -10,41 +10,35 @@ export function createConcurrentSafePromise() {
   let basePromiseId = -1;
   let latestResolvedId = -1;
   let latestResolvedValue: unknown = undefined;
-  let runningPromisesCount = 0;
 
-  function runConcurrentSafePromise<TValue>(promise: MaybePromise<TValue>) {
+  return function runConcurrentSafePromise<TValue>(
+    promise: MaybePromise<TValue>
+  ) {
     basePromiseId++;
-    runningPromisesCount++;
     const currentPromiseId = basePromiseId;
 
-    return Promise.resolve(promise)
-      .then((x) => {
-        // The promise might take too long to resolve and get outdated. This would
-        // result in resolving stale values.
-        // When this happens, we ignore the promise value and return the one
-        // coming from the latest resolved value.
-        //
-        // +----------------------------------+
-        // |        100ms                     |
-        // | run(1) +--->  R1                 |
-        // |        300ms                     |
-        // | run(2) +-------------> R2 (SKIP) |
-        // |        200ms                     |
-        // | run(3) +--------> R3             |
-        // +----------------------------------+
-        if (latestResolvedValue && currentPromiseId < latestResolvedId) {
-          return latestResolvedValue as TValue;
-        }
+    return Promise.resolve(promise).then((x) => {
+      // The promise might take too long to resolve and get outdated. This would
+      // result in resolving stale values.
+      // When this happens, we ignore the promise value and return the one
+      // coming from the latest resolved value.
+      //
+      // +----------------------------------+
+      // |        100ms                     |
+      // | run(1) +--->  R1                 |
+      // |        300ms                     |
+      // | run(2) +-------------> R2 (SKIP) |
+      // |        200ms                     |
+      // | run(3) +--------> R3             |
+      // +----------------------------------+
+      if (latestResolvedValue && currentPromiseId < latestResolvedId) {
+        return latestResolvedValue as TValue;
+      }
 
-        latestResolvedId = currentPromiseId;
-        latestResolvedValue = x;
+      latestResolvedId = currentPromiseId;
+      latestResolvedValue = x;
 
-        return x;
-      })
-      .finally(() => runningPromisesCount--);
```

Comment:
I removed the entire mechanism that was introduced in #829 because counting promises is unreliable—we just cancel promises now and check `store.pendingRequests.isEmpty` now, we don't rely on a counter. The `createConcurrentSafe` implementation is essentially back to what it was before the previous fix.

Reply:
we're still counting promises here right? https://github.com/algolia/autocomplete/pull/864/files/5163e7d7635f0c6c0f98e7d70eed86ec4cf6f480#diff-bc6bfb2dd9285049832c07c042b0b53aedd4bf38c455a3dd3746ebdeee767e99R17

Reply:
If you're referring to `basePromiseId`, that's unrelated—we just increment a counter to know which promise is the current one and perform logic to return the freshest value.

The logic to determine whether there are _pending_ promises was done with `runningPromisesCount` and that code is completely gone. We handle this with [`createCancelablePromiseList#isEmpty`](https://github.com/algolia/autocomplete/blob/d68de50779d5f39bb645d5da092ad3ab1c7fc471/packages/autocomplete-core/src/utils/createCancelablePromiseList.ts#L27-L29) now.